### ISSUE TITLE
Use localStorage shim in certain private browsers

### ISSUE THREAD

**Issue by petermikitsh** [feathersjs/feathers-authentication#135]
The [getStorage implementation](https://github.com/feathersjs/feathers-authentication/blob/master/src/client/utils.js#L67-L81) needs to be augmented to supported private browsers on Safari OS X, Safari iOS, and Android browsers.

Users typically supply their own storage implementation:

```js
import authentication from 'feathers-authentication/client'
import feathers from 'feathers/client'

var client = feathers()
        .configure(authentication({storage: window.localStorage}));
```

`window.localStorage` is defined on these browsers, so it will be returned, rather than the shim. Consequently, any calls to `setItem` (e.g., https://github.com/feathersjs/feathers-authentication/blob/fca25f2b5b03b773975595ae9ac034199df5127b/src/client/index.js#L40-L42) will throw exceptions.

More info [in this comment](https://github.com/feathersjs/feathers-authentication/issues/132#issuecomment-202599102) on #132.

**Comment by daffl** [comment]
The crappy part in Safari is that you actually have to call `setItem` to know if it is working or not. I guess we have to `setItem` a random string, catch the exception and return the in-memory implementation.

**Comment by petermikitsh** [comment]
Yup, that's what [meteor does](https://github.com/meteor/meteor/blob/dc3cd6eb92f2bdd1bb44000cdd6abd1e5d0285b1/packages/localstorage/localstorage.js#L8-L18).

It's not elegant, but it's certainly an accepted solution to resolving the corner case.

**Comment by marshallswain** [comment]
I'd like to help get this implemented.  Instead of using the memory provider, why don't we use cookies?  They work in Private Browsing mode, and as long as we're not consuming them on the server, they're basically the same as localStorage.  The additional bonus of using them would be that they persist on refreshes, where memory-based storage won't.  Maybe we should make this part configurable: `fallback: `cookie`` by default, but you can specify `fallback: 'memory'` if needed.

**Comment by ekryski** [comment]
I actually think in 95% of use cases it already works just fine. That's just me throwing out a random number and pretending to be definitive.... but how many people actually use private browsing mode only on Safari? It sort of defeats the purpose of storing any token anyway in private browsing mode... I think the shim is a totally valid option. I'm also wondering if it is something that should just be left up to the developer to implement??

You have to think about React Native as well, which doesn't support cookies unless you are inside a webview because RN isn't a browser.

**Comment by marshallswain** [comment]
The number of users you encounter using Safari in Private Browsing mode increases a lot when you are talking about a financial application, or any app where users naturally want to do more to protect potentially-sensitive data.  In this case, though, the app can simply pass a cookie storage engine and not worry about localStorage at all.  I can always check if localStorage is available and switch the storageEngine based on that.

```js
function getTokenLocation(){
  let localStorageAvailable = true;

  // Only run this in the browser. (localStorage isn't available on the SSR server)
  if (window.localStorage) {
    // Try localStorage
    try{
      localStorage.setItem('2398sdvlkn4v', 'test');
      localStorage.removeItem('2398sdvlkn4v');
    }
    catch(err){
      localStorageAvailable = false;
    }

    // If localStorage isn't available, try reading from a cookie.
    if (localStorageAvailable) {
      return 'localStorage';
    } else {
      // document.cookie = 'feathers-jwt=hi';
      return 'cookie';
    }
  }
}
getTokenLocation();
```

**Comment by marshallswain** [comment]
Looks like somebody already made a web storage compatible syntax around document.cookie:

https://www.npmjs.com/package/cookie-storage

**Comment by marshallswain** [comment]
Why don't we get opinionated about the storage engine used by default?  We can make the user experience better by providing a `useStorage` boolean in the options.  If the dev also provides the `storage` option, then we use what they provide, otherwise we use one of the following:

1. localStorage - if available
2. cookieStorage
3. somethingStorage - whatever React Native needs.

**Comment by marshallswain** [comment]
If we want to keep it more modular and not have the cookie-storage package as part of this one, then maybe we just allow the `storage` option to take an array and the `authenticate` method only has to check at each step to see if it stored properly.

**Comment by daffl** [comment]
You can easily provide your own store by passing an object that implements an interface like:

```js
 {
    getItem(key) {},
    setItem(key, value) {}
  }
```

We can add a `defaultStorage` fallback that you can set if the check discussed above fails as part of the fix for this issue.

**Comment by marshallswain** [comment]
@ekryski what works for React Native?

**Comment by petermikitsh** [comment]
I believe this issue should be closed and a new issue should be created under feathers-authentication-client.

**Comment by marshallswain** [comment]
I think this can be solved by directing people to use localForage. We can focus on solutions that others aren't already solving.  Anybody object to closing this?

**Comment by petermikitsh** [comment]
Here's my current solution for safari (private browsing mode):

```
/* eslint no-undefined: 0 */

export default function getStorage() {
  const storageKey = '_localstorage_test_';
  let retrieved;
  try {
    if (window.localStorage) {
      window.localStorage.setItem(storageKey, storageKey);
      retrieved = window.localStorage.getItem(storageKey);
      window.localStorage.removeItem(storageKey);
    }
  } catch (e) {
    // ... ignore
  }

  if (storageKey === retrieved) {
    return window.localStorage;
  }

  return {
    data: {},
    setItem: function (key, val) {
      this.data[key] = val;
    },
    removeItem: function (key) {
      delete this.data[key];
    },
    getItem: function (key) {
      const value = this.data[key];
      return value === undefined ? null : value;
    }
  };
}
```

init like this:

```
import auth from 'feathers-authentication-client';
import feathers from 'feathers/client';
import hooks from 'feathers-hooks';
import io from 'socket.io-client';
import socketio from 'feathers-socketio/client';

const app = feathers()
  .configure(hooks())
  .configure(socketio(socket))
  .configure(auth({
    storage: getStorage()
  }));
```

**Comment by petermikitsh** [comment]
Yeah @marshallswain I agree that for web the use case, we should probably just be a documentation  update to discuss edge cases when using localStorage (in normal web environments), an inform people to use one of the polyfills in environments where `window.localStorage` acts weird, like Safari with Private Browsing.

I believe for React Native you'd probably just use RN's [AsyncStorage](https://facebook.github.io/react-native/docs/asyncstorage.html), **however** on RN you're given a Promise-based API for `getItem` and `setItem` (unlike `window.localStorage`, which is synchronous).

```js
import auth from 'feathers-authentication-client';
import feathers from 'feathers/client';
import hooks from 'feathers-hooks';
import io from 'socket.io-client';
import socketio from 'feathers-socketio/client';
import {AsyncStorage} from 'react-native';

const socket = io('http://api/url');
const app = feathers()
  .configure(hooks())
  .configure(socketio(socket))
  .configure(auth({
    storage: AsyncStorage
  }));
```

In light of that, you may want to ensure that the storage adapter could work with Promise-based API. I'm not sure if it does this currently. But it wouldn't be hard to add support for it.

**Comment by corymsmith** [comment]
This works with AsyncStorage as @ekryski and I are using it in a few RN apps
